app/services/middleware_topology_service.rb

```diff
@@ -33,14 +36,14 @@ def entity_display_type(entity)
 
   def build_entity_data(entity)
     data = build_base_entity_data(entity)
-    data.merge!(:status => 'Unknown',
-                :display_kind => entity_display_type(entity))
-    data[:icon] = entity.decorate.try(:listicon_image)
+    data[:status] = 'Unknown'
+    data[:display_kind] = entity_display_type(entity)
+    data[:icon] = entity.decorate.try(:listicon_image) unless entity.kind_of?(MiddlewareDatasource)
```

Comment:
Although I like your proposal, it's not easily doable, the topology graph is an svg image and it doesn't support the CSS pseudo elements (:before :after) https://stackoverflow.com/questions/14938201/unexpected-duplication-of-after-pseudo-element-assigned-to-svgtext-working-wi

That's actually why the glyphs are stored in the code as unicode characters and not as thereadable `fa fa-something` css class "selectors"

btw. I rewrote it to this, but then figured out, that it can work only if the decorator will return the unicode char directly and not the `fa fa-database`, but this will break the other occurences that depend on this form. 
```ruby
    # fonticon takes precedence before the listicon_image
    data[:icon] = if icon = entity.decorate.try(:fonticon)
                    {
                      :type  => :glyph,
                      :value => icon
                    }
                  elsif icon = entity.decorate.try(:listicon_image)
                    {
                      :type  => :image,
                      :value => "/assets/svg/#{icon}.svg"
                    }
                  end
    data
```

I'd stick with the `UiServiceMixin`, 1 global or split it into smaller ones. Both approaches have it's own benefits. IMO I am spending too much time on this :)